backend/src/worker.rs

```diff
@@ -1351,175 +1361,228 @@ async fn handle_child(
+    let wait_db = db.clone();
+    let wait = async move {
+        let db = wait_db;
+
+        let timed_out = tokio::select! {
+            biased;
+            result = child.wait() => return result.map(Some),
+            _ = too_many_logs.changed() => false,
+            _ = cancel_check => false,
+            _ = sleep(timeout) => true,
+        };
+
+        let set_reason = async {
+            if timed_out {
+                if let Err(err) = sqlx::query(
+                    r#"
+                       UPDATE queue
+                          SET canceled = true
+                            , canceled_by = 'timeout',
+                            , canceled_reason = $1
+                        WHERE id = $2
+                    r"#,
+                )
+                .bind(format!("duration > {}", timeout.as_secs()))
+                .bind(job_id)
+                .execute(&db)
+                .await
+                {
+                    tracing::error!(%job_id, %err, "error setting cancelation reason");
                 }
-            } => {
-                child.kill().await?;
-                return Err(Error::ExecutionErr("execution interrupted".to_string()).into())
             }
         };
-        r
-    });
-
-    let (tx, mut rx) = mpsc::channel::<String>(100);
-
-    tokio::spawn(async move {
-        while !done4.load(Ordering::Relaxed) {
-            let send = tokio::select! {
-                Ok(Some(out)) = reader.next_line() => {
-                    if out.len() > MAX_LOG_SIZE as usize {
-                        tracing::info!("Line is too big");
-                        let _ = tx.send(format!("Line is too big")).await;
-                        done4.store(true, Ordering::Relaxed);
-                        break;
-                    } else {
-                        tx.send(out).await
+
+        /* send SIGKILL and reap child process */
+        let (_, kill) = futures::future::join(set_reason, child.kill()).await;
+        kill.map(|()| None)
+    };
+
+    /* a future that reads output from the child and appends to the database */
+    let lines = async move {
+        /* log_remaining is zero when output limit was reached */
+        let mut log_remaining = (MAX_LOG_SIZE as usize).saturating_sub(logs.chars().count());
+        let mut result = io::Result::Ok(());
+        let mut output = output;
+        let mut write: OptionFuture<_> = None.into();
+
+        while let Some(line) = output.by_ref().next().await {
+            /* after receiving a line, continue until some delay has passed
+             * _and_ the previous database write is complete */
+            let until_write = write.clone();
+            let until = sleep(write_logs_delay).then(|()| until_write);
+            let mut read_lines = ready(line)
+                .into_stream()
+                .chain(output.by_ref())
+                .take_until(until)
+                .boxed();
```

Comment:
that was a bit hard to read, extract into a function ?

read_lines_until(duration) ?